### PR TITLE
[PROTOTYPE] Support Transpose DataFrame API

### PR DESCRIPTION
**Pull request by xinrong-meng**
Prototype

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Alias, Ascending, Attribute, AttributeReference, Cast, SortOrder}
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan, Project, Sort, Transpose}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern
+import org.apache.spark.sql.types.{AtomicType, DataType, StringType}
+import org.apache.spark.unsafe.types.UTF8String
+
+
+class ResolveTranspose(sparkSession: SparkSession) extends Rule[LogicalPlan] {
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala:30]: I guess it should take a `val catalogManager: CatalogManager` as parameter?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2175,6 +2175,105 @@ class Dataset[T] private[sql](
...
+   *
+   * @param indexColumn
+   *   The single column that will be treated as the index for the transpose operation.This column
+   *   will be used to pivot the data, transforming the DataFrame such that the values of the
+   *   indexColumn become the new columns in the transposed DataFrame.
+   * @group untypedrel
+   * @since 4.0.0
+   */
+  def transpose(indexColumn: Column): DataFrame = withPlan {
+    if (this.isEmpty && this.columns.isEmpty) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: we should not do this early optimization. `this.isEmpty` will trigger a job.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(TreePattern.UNRESOLVED_TRANSPOSE)) {
+    case t @ UnresolvedTranspose(indexColumn, child) if indexColumn.resolved && child.resolved =>
+
+      // Cast the index column to StringType
+      val indexColumnAsString = indexColumn match {
+        case attr: Attribute if attr.dataType.isInstanceOf[AtomicType] =>
+          Alias(Cast(attr, StringType), attr.name)()
+        case attr: Attribute =>
+          throw new IllegalArgumentException(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: it's a user facing error, we should have an error class.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+      val fullCollectedRows = queryExecution.executedPlan.executeCollect()
+
+      if (fullCollectedRows.isEmpty) {
+        // Return a DataFrame with a single column "key" containing non-index column names
+        val keyAttr = AttributeReference("key", StringType, nullable = false)()
+        val keyValues = nonIndexColumnsAttr.map(
+          _.name).map(name => UTF8String.fromString(name))
+        val keyRows = keyValues.map(value => InternalRow(value))
+
+        LocalRelation(Seq(keyAttr), keyRows)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala:145]: can we still return `Transpose` here?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      // Sort by index column values, and collect the casted frame
+      val allCastCols = indexColumnAsString +: nonIndexColumnsAsLCT
+      val sortedChild = Sort(
+        Seq(SortOrder(indexColumn.asInstanceOf[Attribute], Ascending)),
+        global = true,
+        child
+      )
+      val projectAllCastCols = Project(allCastCols, sortedChild)
+      val queryExecution = sparkSession.sessionState.executePlan(projectAllCastCols)
+      val fullCollectedRows = queryExecution.executedPlan.executeCollect()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala:136]: shall we add a LIMIT to avoid collecting too many data?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala:136]: we can add LIMIT dataFrameTransposeMaxValues + 1

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        val keyAttr = AttributeReference("key", StringType, nullable = false)()
+        val keyValues = nonIndexColumnsAttr.map(
+          _.name).map(name => UTF8String.fromString(name))
+        val keyRows = keyValues.map(value => InternalRow(value))
+
+        LocalRelation(Seq(keyAttr), keyRows)
+      } else {
+        val maxValues = sparkSession.sessionState.conf.dataFrameTransposeMaxValues
+        if (fullCollectedRows.length > maxValues) {
+          throw new IllegalArgumentException(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: ditto, need an error class

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Makes sense removed.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala (hunk 4 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala:145]: Good catch! Adjusted

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Can we keep `this.columns.isEmpty`? I realized `this.columns.isEmpty` indicates `this.isEmpty` implicitly.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala (hunk 3 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: Cool

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala (hunk 5 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala:136]: Good idea. Added.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala (hunk 6 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: Cool

Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala (hunk 3 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: Done

**Comment by xinrong-meng** [comment]
I filed a PR against OSS, for further comments/reviews please use the OSS PR, thank you!

**Pull request closed without merging**